### ISSUE TITLE
Generic `unions`

### ISSUE THREAD

**Issue by JustusAdam** [tibbe/unordered-containers#185]
Since `base` now uses `Foldable` basically everywhere would it not be a good idea to change `unions` for both `HashMap`s and `HashSet`s to `Foldable f` rather than `[]`?

Since the implementation is `foldl'` anyways there should be no performance difference.

**Comment by treeowl** [comment]
Yes. I'd really rather get rid of `unions` and `unionsWith`, but a recent discussion on the libraries list shows that a number of people are quite attached to them.

**Comment by JustusAdam** [comment]
Well, it turns out to be a rather common use case. I personally use it a lot in tree traversals for aggregating information from subtrees. However my trees are all `Foldable` or `Traversable`, as such it would be nice to have a `Foldable` constrained `unions` so I could call them directly on the tree rather than having to do `unions . toList`. Its just a little nicer

**Comment by sjakobi** [comment]
`unions` in `containers` takes `Foldable`, so we should probably do the same.

**Comment by treeowl** [comment]
Sure... Should both use foldMap'?

**Comment by sjakobi** [comment]
> Sure... Should both use foldMap'?

If that would preserve the current semantics, yes.

**Comment by treeowl** [comment]
We'd have to be careful to use the right order, but it should be fine.
Something like this, I *think*:

newtype Onion k v = Onion { scrunion :: HashMap k v }

instance (Eq k, Hashable k) => Semigroup (Onion k v) where
  -- this is the spot we have to check for ordering
  Onion t <> Onion u = Onion (union u t)
instance (Eq k, Hashable k) => Monoid (Onion k v) where
  mempty = Onion empty

unions = scrunion . foldMap' Onion

We'll have to check that performance is ever better than just using foldl'.
